File: 백준/Gold/17070. 파이프 옮기기 1/파이프 옮기기 1.py
# 파이프 옮기기1
import sys
input = sys.stdin.readline

def backtrack(i, j, pos):

    if i == n-1 and j == n-1: # 끝에 도착했다면 1을 리턴
        return 1

    if dp[pos][i][j] != -1:
        return dp[pos][i][j]
    else:
        dp[pos][i][j] = 0

    for dx, dy, nPos in dic[pos]:
        nx, ny = i + dx, j + dy
   
        if nx < 0 or nx >= n or ny < 0 or ny >= n:
            continue
        elif graph[nx][ny] == 1:
            continue

        if nPos == 2: # 대각선이동은 총 3칸을 확인해야함
            if graph[i+1][j] == 1 or graph[i][j+1] == 1:
                continue

        dp[pos][i][j] += backtrack(nx, ny, nPos)
        
    return dp[pos][i][j]

n = int(input())
graph = []

for i in range(n):
    graph.append(list(map(int,input().split())))
    
dp = [[[-1] * n for _ in range(n)] for _ in range(3)]
dic = {0: [(0, 1, 0), (1, 1, 2)], 
       1:[(1, 0, 1), (1, 1, 2)],
       2:[(0,1, 0), (1, 0, 1), (1, 1, 2)]}

backtrack(0, 1, 0)

print(dp[0][0][1])

File: 백준/Gold/17070. 파이프 옮기기 1/파이프 옮기기 1.java
import java.io.*;
import java.util.*;

public class Main {
    static int[][] graph;
    static int n;
    static int[][][] dp;
    static HashMap<Integer, int[][]> posMap;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        n = Integer.parseInt(br.readLine());
        graph = new int[n][n];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                graph[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        dp = new int[3][n][n];

        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < n; j++) {
                for (int k = 0; k < n; k++) {
                    dp[i][j][k] = -1;
                }
            }
        }
        posMap = new HashMap<>(n);
        posMap.put(0, new int[][] {{0,1,0}, {1,1,2}});
        posMap.put(1, new int[][] {{1,0,1}, {1,1,2}});
        posMap.put(2, new int[][] {{0,1,0}, {1,0,1}, {1,1,2}});

        backtrack(0, 1, 0);

        System.out.println(dp[0][0][1]);
    }

    static int backtrack(int i, int j, int pos) {
        if (i == n-1 && j == n-1) {
            return 1;
        }

        if (dp[pos][i][j] != -1) {
            return dp[pos][i][j];
        } else {
            dp[pos][i][j] = 0;
        }

        for (int[] next : posMap.get(pos)) {
            int dx = next[0];
            int dy = next[1];
            int nPos = next[2];

            int nx = i + dx;
            int ny = j + dy;

            if (nx < 0 || ny < 0 || nx >= n || ny >= n) {
                continue;
            } else if (graph[nx][ny] == 1) {
                continue;
            }

            if (nPos == 2) {
                if (graph[i+1][j] == 1 || graph[i][j+1] == 1) {
                    continue;
                }
            }
            dp[pos][i][j] += backtrack(nx, ny, nPos);
        }
        return dp[pos][i][j];
    }
}
